blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/AbstractElasticIO.java

```diff
@@ -0,0 +1,444 @@
+    public static String ELASTICSEARCH_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_WRITE);
+    public static String ELASTICSEARCH_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_INDEX_NAME_READ);
+
+    public static String ENUMS_INDEX_NAME_WRITE = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_WRITE);
+    public static String ENUMS_INDEX_NAME_READ = Configuration.getInstance().getStringProperty(ElasticIOConfig.ELASTICSEARCH_ENUMS_INDEX_NAME_READ);
+
+    private final int LEVEL_0 = 0;
+    private final int LEVEL_1 = 1;
+    private final String REGEX_TOKEN_DELIMTER = "\\.";
+
+    public List<SearchResult> search(String tenant, String query) throws Exception {
+        return search(tenant, Arrays.asList(query));
+    }
+
+
+    public List<SearchResult> search(String tenant, List<String> queries) throws Exception {
+        String[] indexes = getIndexesToSearch();
+
+        return searchByIndexes(tenant, queries, indexes);
+    }
+
+    private List<SearchResult> searchByIndexes(String tenant, List<String> queries, String[] indexes) {
+        List<SearchResult> results = new ArrayList<SearchResult>();
+        Timer.Context multiSearchCtx = searchTimer.time();
+        queryBatchHistogram.update(queries.size());
+        BoolQueryBuilder bqb = boolQuery();
+        QueryBuilder qb;
+
+        for (String query : queries) {
+            GlobPattern pattern = new GlobPattern(query);
+            if (!pattern.hasWildcard()) {
+                qb = termQuery(ESFieldLabel.metric_name.name(), query);
+            } else {
+                qb = regexpQuery(ESFieldLabel.metric_name.name(), pattern.compiled().toString());
+            }
+
+            bqb.should(boolQuery()
+                            .must(termQuery(ESFieldLabel.tenantId.toString(), tenant))
+                            .must(qb)
+            );
+        }
+
+        SearchResponse response = client.prepareSearch(indexes)
+                .setRouting(tenant)
+                .setSize(100000)
+                .setVersion(true)
+                .setQuery(bqb)
+                .execute()
+                .actionGet();
+
+        multiSearchCtx.stop();
+
+        for (SearchHit hit : response.getHits().getHits()) {
+            SearchResult result = convertHitToMetricDiscoveryResult(hit);
+            results.add(result);
+        }
+        return dedupResults(results);
+    }
+
+    /**
+     * If we think of metric name(Ex: foo.bar.baz) as hierarchical tokens separated by delimiter("."),
+     * this API method would return the next level of tokens which follow the given prefix.
+     *
+     * @param tenant
+     * @param prefix is glob representation
+     * @return
+     * @throws Exception
+     */
+    public List<TokenInfo> getNextTokens(final String tenant, final String prefix) throws Exception {
```

Comment:
We'll see what others say, but might make it easier to understand if you broken this method down a bit more.  It would make it easier to glance at and see the different parts of the process.